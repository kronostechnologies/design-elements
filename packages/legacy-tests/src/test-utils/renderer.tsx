import { CommonWrapper, mount, MountRendererProps, ReactWrapper, render } from 'enzyme';
import { act, Component, FunctionComponent, PropsWithChildren, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DesignSystem, DesignSystemProps } from '~/components/design-system';
import { ThemeWrapper } from '~/components/theme-wrapper/theme-wrapper';

export const AllProviders: FunctionComponent<PropsWithChildren<DesignSystemProps>> = ({
    children,
    language,
    theme,
    staticDevice,
}) => (
    <MemoryRouter future={{ v7_startTransition: false, v7_relativeSplatPath: false }}>
        <DesignSystem language={language} staticDevice={staticDevice} theme={theme}>
            {children}
        </DesignSystem>
    </MemoryRouter>
);

interface WrappingComponentProps {
    wrappingComponentProps?: DesignSystemProps;
}

type Options = MountRendererProps & WrappingComponentProps;

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options: Options = {},
): ReactWrapper<P, S, C> {
    return mount(component, {
        ...options,
        wrappingComponent: AllProviders as Options['wrappingComponent'],
    });
}

export function mountWithTheme<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options?: MountRendererProps,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: ThemeWrapper as MountRendererProps['wrappingComponent'],
        ...options,
    });
}

export function renderWithTheme(
    component: ReactElement,
): cheerio.Cheerio {
    return render(
        <ThemeWrapper>
            {component}
        </ThemeWrapper>,
    );
}

export async function actUpdate<C extends Component, P = C['props'], S = C['state']>(
    wrapper: CommonWrapper<P, S, C>,
): Promise<void> {
    await act(async () => {
        wrapper.update();
    });
}

export async function actAndWaitForEffects<C extends Component, P = C['props'], S = C['state']>(
    wrapper: ReactWrapper<P, S, C>,
    action: () => void,
): Promise<ReactWrapper<P, S, C>> {
    await act(async () => action());
    await actUpdate(wrapper);
    return wrapper;
}
