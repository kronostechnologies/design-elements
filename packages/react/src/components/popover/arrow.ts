import styled from 'styled-components';

const arrowSize = 24;
const height = arrowSize / 3;

export const Arrow = styled.div`
    height: ${arrowSize}px;
    position: absolute;
    width: ${arrowSize}px;

    &[data-placement*="bottom"] {
        height: ${height}px;
        left: 0;
        margin-top: -${height - 2}px;
        top: 0;
        width: ${arrowSize}px;

        &::before {
            border-color: transparent transparent #fff transparent;
            border-width: 0 ${arrowSize / 2}px ${height}px ${arrowSize / 2}px;
        }
    }

    &[data-placement*="top"] {
        bottom: 0;
        height: ${height}px;
        left: 0;
        margin-bottom: -${height - 2}px;
        width: ${arrowSize / 2}px;

        &::before {
            border-color: #fff transparent transparent transparent;
            border-width: ${height}px ${arrowSize / 2}px 0 ${arrowSize / 2}px;
        }
    }

    &[data-placement*="right"] {
        height: ${arrowSize}px;
        left: 0;
        margin-left: -${height - 2}px;
        width: ${height}px;

        &::before {
            border-color: transparent #fff transparent transparent;
            border-width: ${arrowSize / 2}px ${height}px ${arrowSize / 2}px 0;
        }
    }

    &[data-placement*="left"] {
        height: ${arrowSize}px;
        margin-right: -${height - 2}px;
        right: 0;
        width: ${height}px;

        &::before {
            border-color: transparent transparent transparent #fff;
            border-width: ${arrowSize / 2}px 0 ${arrowSize / 2}px ${height}px;
        }
    }

    &::before {
        border-style: solid;
        content: "";
        display: block;
        height: 0;
        margin: auto;
        width: 0;
    }
`;
