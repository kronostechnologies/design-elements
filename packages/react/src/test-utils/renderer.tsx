import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
