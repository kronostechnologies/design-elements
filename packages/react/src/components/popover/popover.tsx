import React, { ReactElement, ReactNode, useEffect, useRef, useState } from 'react';
import { Manager, Popper, PopperChildrenProps, PopperProps, Reference, ReferenceChildrenProps } from 'react-popper';

import ReactDOM from 'react-dom';
import FocusLock from 'react-focus-lock';
import styled from 'styled-components';
import { Arrow } from './arrow';

const Wrapper = styled.div<{ distance: number }>`
    background-color: #fff;
    border-radius: 4px;
    filter: drop-shadow(0 0 10px rgba(0, 0, 0, 0.18));
    padding: 8px;
    position: relative;

    &[data-placement*="bottom"] {
        margin-top: ${({ distance }) => distance}px;
    }

    &[data-placement*="top"] {
        margin-bottom: ${({ distance }) => distance}px;
    }

    &[data-placement*="right"] {
        margin-left: ${({ distance }) => distance}px;
    }

    &[data-placement*="left"] {
        margin-right: ${({ distance }) => distance}px;
    }
`;

interface PopoverProps {
    // The popover's content
    children: ReactNode;
    // Target on which the popover is attached
    target: ReactElement;
    /**
     * Position relative to the target
     * @default auto
     */
    position?: PopperProps['placement'];
    /**
     * Distance in px between the target and the popover
     * @default 8
     */
    distance?: number;
    /**
     * Display an arrow with the popover
     * @default true
     */
    arrow?: boolean;
}

export function Popover({
    children,
    target,
    position = 'auto',
    distance = 8,
    arrow = true,
}: PopoverProps): ReactElement {
    const targetRef = useRef<HTMLElement | null>(null);
    const popoverRef = useRef<HTMLElement | null>(null);

    const [isOpen, setIsOpen] = useState(false);
    const [isLockDisabled, setIsLockDisabled] = useState(true);
    const [shouldRestoreFocus, setShouldRestoreFocus] = useState(false);

    function togglePopover(): void {
        setIsOpen(prevState => !prevState);
    }

    function releaseFocusLock(): void {
        setIsLockDisabled(true);
        setShouldRestoreFocus(true);
        setIsOpen(false);
    }

    // Position the focus onto the popover on opening
    useEffect(() => {
        if (isOpen && popoverRef.current) {
            popoverRef.current.focus();
        }
    }, [isOpen]);

    // Restore focus on target element
    useEffect(() => {
        if (shouldRestoreFocus && targetRef.current) {
            targetRef.current.focus();
        }

        return () => {
            setShouldRestoreFocus(false);
        };
    }, [shouldRestoreFocus]);

    // Close popover with keyboard escape key
    useEffect(() => {
        function handleKeypress(e: KeyboardEvent): void {
            if (e.key === 'Escape') {
                releaseFocusLock();
            }
        }

        window.addEventListener('keydown', handleKeypress);

        return () => {
            window.removeEventListener('keydown', handleKeypress);
        };
    }, []);

    useEffect(
        () => {
            function listener(event: Event): void {
                const { current: targetEl } = targetRef;
                const { current: popoverEl } = popoverRef;

                // Do nothing on popoverRef's element click or descendent elements, or clicking targetRef's element
                if (!popoverEl || popoverEl.contains(event.target as Node) ||
                    !targetEl || targetEl.contains(event.target as Node)) {
                    return;
                }

                releaseFocusLock();
            }

            document.addEventListener('mousedown', listener);
            document.addEventListener('touchstart', listener);

            return () => {
                document.removeEventListener('mousedown', listener);
                document.removeEventListener('touchstart', listener);
            };
        },
        [targetRef, popoverRef, releaseFocusLock],
    );

    function renderTarget(referenceProps: ReferenceChildrenProps): ReactElement {
        const targetClassName = target.props.className || '';

        const targetProps = {
            ...target.props,
            ref: referenceProps.ref,
            onClick: togglePopover,
            className: isOpen ? `${targetClassName} active` : targetClassName,
        };

        return React.cloneElement(target, targetProps);
    }

    return (
        <Manager>
            <Reference innerRef={ref => targetRef.current = ref}>{renderTarget}</Reference>
            {isOpen &&
                ReactDOM.createPortal(
                    <Popper placement={position} innerRef={ref => popoverRef.current = ref}>
                        {({ ref, style, placement, arrowProps }: PopperChildrenProps) => (
                            <Wrapper
                                ref={ref}
                                style={style}
                                data-placement={placement}
                                tabIndex={0}
                                distance={distance}
                            >
                                <FocusLock disabled={isLockDisabled}>
                                    {children}
                                    <div onFocus={releaseFocusLock} tabIndex={0} />
                                </FocusLock>
                                {arrow &&
                                    <Arrow ref={arrowProps.ref} style={arrowProps.style} data-placement={placement} />
                                }
                            </Wrapper>
                        )}
                    </Popper>,
                    document.body,
                )}
        </Manager>
    );
}
