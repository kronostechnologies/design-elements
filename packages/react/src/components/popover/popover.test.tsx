import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import { Popover } from './popover';

describe('Popover', () => {
    test('Matches the snapshot', () => {
        const tree = renderer.create(
            <Popover target={<button>target</button>}>
                <div>content</div>
            </Popover>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('Toggle the popover content into the dom on target click', () => {
        const wrapper = mount(
            <Popover target={<button id="target">target</button>}>
                <div id="content">popover content</div>
            </Popover>,
        );

        const targetEl = wrapper.find('#target');

        targetEl.simulate('click');
        expect(document.getElementById('content')).not.toBeNull();

        targetEl.simulate('click');
        expect(document.getElementById('content')).toBeNull();
    });
});
