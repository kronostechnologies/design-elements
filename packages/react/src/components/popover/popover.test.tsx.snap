// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Matches the snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
>
  target
</button>
`;
